node/src/reactor/joiner.rs

```diff
@@ -943,7 +944,7 @@ impl Reactor {
             contract_runtime: self.contract_runtime,
             storage: self.storage,
             consensus: self.consensus,
-            latest_block: self.linear_chain.latest_block().clone(),
+            latest_block: self.linear_chain_sync.latest_block().cloned(),
```

Comment:
I disagree with this change – it should still come from the `linear_chain` and if you need it to be something else than _just_ the last block (for example with changed applied) then it should also be captured there.

Reply:
The problem is, `LinearChain` only updates its `latest_block` when it handles a block that's executed and passed through consensus. And it turns out that it's possible that it will handle no block at all - right after the upgrade, when the trusted hash is the hash of the switch block before the upgrade. Then no block will be executed, and so the `LinearChain` will _not_ update that field, which would give `None` here. That led to the `BlockProposer` being initialized with `next_finalized = 0`, which made nodes unable to propose blocks.

I did it this way, because `LinearChainSync` is also aware of the latest block after syncing, and it is _always_ aware of it, even if no block was executed. Then the correct value is passed to the `BlockProposer` after an upgrade and everything works.

Reply:
Then make  `LinearChain` always aware of the last block. That's the component that should every other component consult about the state of the linear chain.

Reply:
But I guess it's yet another example of working under pressure, cutting corners, creating tech debt and making sacrifices.